### PR TITLE
[reggen] Further changes, replacing loaded data with a proper class

### PR DESCRIPTION
**Pull request by rswarbrick**
This finishes up the campaign to replace OrderedDicts in reggen with actual classes. The four commits:
1. Define a nice params interface
1. Represent interrupts and alerts with a "Signal" type
1. Define a type for inter-module signals
1. (Finally) wrap the whole block up in a class, replacing the top-level dict.

### CODE REVIEW

**Review by msfschaffner — APPROVED** [review]
Thanks for the extensive refactoring @rswarbrick
I looked through the change and it LGTM. 
I like the consolidation this provides.

However, since it is a pretty large change, it would be good to get a second pair of eyes on it.

**Comment by rswarbrick** [comment]
This PR is going for a prize for "most rebased" because it touches the checked-in autogenerated code and conflicts with pretty much *any* change to the top-level. Eugh. Anyway, rebased again...

File: util/reggen/block.py [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+        's', "License ientifier (if using pure json) "
+        "Only use this if unable to put this "
+        "information in a comment at the top of the "
+        "file."
+    ],
+    'wakeup_list': ['lnw', "list of peripheral wakeups"]
+}
+
+
+class Block(BlockBase):
```
Comment (imphil) [inline util/reggen/block.py:74]: The thing represented here is typically called IP or IP block in OpenTitan -- block on its own is rarely used. Can you rename the class e.g. to `IpBlock`, and maybe the base class to Block?

File: util/reggen/chip.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+from .block_base import BlockBase
+from .block import Block
+from .params import Params
+from .reg_block import RegBlock
+from .window import Window
+
+_Triple = Tuple[int, str, Block]
+
+
+class Chip(BlockBase):
```
Comment (imphil) [inline util/reggen/chip.py:18]: In OpenTitan, the thing represented by this class is typically called a Toplevel. Not that this name is necessarily better, but it's much more common (e.g. "topgen" vs "chipgen", "top_earlgrey" vs "chip_earlgrey" etc). I'd propose renaming this class.

**Review by imphil — APPROVED** [review]
Nice cleanup! Some comments are inline, but overall LGTM.

General note: we now require Python 3.6 which supports f strings. They could be useful for some of the formatting you're doing to cut down the typing in `.format()`. (No need to apply that to this PR, though.)

File: util/reggen/params.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import re
+from typing import Dict, List, Optional
+
+from .lib import check_keys, check_str, check_int, check_bool, check_list
+
+REQUIRED_FIELDS = {
```
[the remark this answers is not in the document]
Comment (rswarbrick) [reply util/reggen/params.py:10]: Hmm. This appears in all the reggen modules and gets inlined into the [generated documentation](https://docs.opentitan.org/doc/rm/register_tool/). I'm hoping that one day we'll have better documentation (more information than the "g" key shows, for example), but I was rather punting on that for now. What sort of short comment do you think would be reasonable?

[the remark this answers is not in the document]
Comment (imphil) [reply util/reggen/params.py:10]: Let's not worry abut it for params.py -- I just realized this syntax is used all over the place and I'm guessing "pb" means "Python Bool" .

File: util/reggen/params.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        if expose:
+            raise ValueError('At {}, the localparam {} cannot be exposed to '
+                             'the top-level.'
+                             .format(where, name))
+        return LocalParam(name, desc, param_type, value=default)
+    else:
+        return Parameter(name, desc, param_type, default, expose)
+
+
+class Params:
```
[the remark this answers is not in the document]
Comment (rswarbrick) [reply util/reggen/params.py:230]: Doing that properly takes [magic involving abstract base classes](https://treyhunner.com/2019/04/why-you-shouldnt-inherit-from-list-and-dict-in-python/), so I thought I wouldn't bother for now. (I *am* getting rid of the list view, as suggested above)

File: util/reggen/block.py (hunk 1 above)
Comment (rswarbrick) [reply util/reggen/block.py]: Done! (Eugh, grepping for that was *hard*...)

File: util/reggen/chip.py (hunk 2 above)
Comment (rswarbrick) [reply util/reggen/chip.py]: Cool, I've gone for `Top` since it's shorter than `Toplevel` and also hasn't got the ambiguous case problem (`TopLevel`?).

**Pull request merged**